File: operator/internal/controller/scoring_test.go
package controller

import (
	"testing"

	baselinev1alpha1 "github.com/maci0/baseline-security-operator/api/v1alpha1"
)

// TestScoringMode: unset and unknown modes fall back to Flat so history and
// status.score use one formula; only SeverityWeighted opts into weights.
func TestScoringMode(t *testing.T) {
	if got := scoringMode(&baselinev1alpha1.ClusterBaseline{}); got != baselinev1alpha1.ScoringFlat {
		t.Fatalf("empty spec = %q, want Flat", got)
	}
	cb := &baselinev1alpha1.ClusterBaseline{}
	cb.Spec.Scoring.Mode = baselinev1alpha1.ScoringFlat
	if got := scoringMode(cb); got != baselinev1alpha1.ScoringFlat {
		t.Fatalf("Flat = %q", got)
	}
	cb.Spec.Scoring.Mode = baselinev1alpha1.ScoringSeverityWeighted
	if got := scoringMode(cb); got != baselinev1alpha1.ScoringSeverityWeighted {
		t.Fatalf("SeverityWeighted = %q", got)
	}
	// Hand-edited / unknown enum value must not invent weights.
	cb.Spec.Scoring.Mode = "UnknownMode"
	if got := scoringMode(cb); got != baselinev1alpha1.ScoringFlat {
		t.Fatalf("unknown mode = %q, want Flat", got)
	}
}

// TestHistoryModeMatches: late CCR refresh must not rewrite history ring points
// under a different formula when the admin flips Flat <-> SeverityWeighted.
// Missing annotation (pre-feature CRs) allows refresh so existing clusters keep
// late-CCR correction; mismatched stamp blocks rewrite.
func TestHistoryModeMatches(t *testing.T) {
	// No annotation: allow refresh (pre-feature CR path).
	cb := &baselinev1alpha1.ClusterBaseline{}
	if !historyModeMatches(cb) {
		t.Fatal("missing annotation must allow late history refresh")
	}
	// Annotation matches current mode.
	cb.Annotations = map[string]string{historyScoringModeAnn: string(baselinev1alpha1.ScoringFlat)}
	if !historyModeMatches(cb) {
		t.Fatal("Flat stamp + Flat mode must match")
	}
	// Mode flipped after history was stamped: block rewrite under new formula.
	cb.Spec.Scoring.Mode = baselinev1alpha1.ScoringSeverityWeighted
	if historyModeMatches(cb) {
		t.Fatal("Flat stamp + SeverityWeighted mode must not match")
	}
	// Stamp matches weighted mode.
	cb.Annotations[historyScoringModeAnn] = string(baselinev1alpha1.ScoringSeverityWeighted)
	if !historyModeMatches(cb) {
		t.Fatal("SeverityWeighted stamp + mode must match")
	}
	// Empty annotation value is treated as missing (allow refresh).
	cb.Annotations[historyScoringModeAnn] = ""
	if !historyModeMatches(cb) {
		t.Fatal("empty stamp must allow refresh")
	}
	// Nil annotations map (distinct from empty string on present map).
	cb.Annotations = nil
	cb.Spec.Scoring.Mode = baselinev1alpha1.ScoringSeverityWeighted
	if !historyModeMatches(cb) {
		t.Fatal("nil annotations must allow refresh even when mode is weighted")
	}
}

func TestScore(t *testing.T) {
	if score(0, 0) != nil || score(-1, 0) != nil || score(1, -1) != nil {
		t.Fatal("zero/negative countable should be nil")
	}
	if s := score(2, 1); s == nil || *s != 66 {
		t.Fatalf("score(2,1) = %v, want 66", s)
	}
	if s := score(1, 0); s == nil || *s != 100 {
		t.Fatalf("score(1,0) = %v, want 100", s)
	}
	if s := score(0, 5); s == nil || *s != 0 {
		t.Fatalf("score(0,5) = %v, want 0", s)
	}
	// Integer division floors: a single FAIL among many PASS must not round up
	// to 100 (would hide the failure on the dashboard).
	if s := score(999, 1); s == nil || *s != 99 {
		t.Fatalf("score(999,1) = %v, want 99 (floor, not round)", s)
	}
	if s := score(1, 2); s == nil || *s != 33 {
		t.Fatalf("score(1,2) = %v, want 33", s)
	}
}

func TestClampScore(t *testing.T) {
	if clampScore(nil) != nil {
		t.Fatal("nil stays nil")
	}
	neg := int32(-1)
	if s := clampScore(&neg); s == nil || *s != 0 {
		t.Fatalf("neg = %v", s)
	}
	hi := int32(101)
	if s := clampScore(&hi); s == nil || *s != 100 {
		t.Fatalf("hi = %v", s)
	}
	ok := int32(77)
	if s := clampScore(&ok); s == nil || *s != 77 {
		t.Fatalf("ok = %v", s)
	}
}

func FuzzScore(f *testing.F) {
	f.Add(int32(0), int32(0))
	f.Add(int32(1), int32(0))
	f.Add(int32(0), int32(1))
	f.Add(int32(2), int32(1))
	f.Add(int32(-1), int32(5))
	f.Add(int32(2147483647), int32(0)) // int32-overflow regression seed
	f.Fuzz(func(t *testing.T, pass, fail int32) {
		s := score(pass, fail)
		// Oracle must use int64 sums (same as score) so int32 overflow is not expected nil.
		if pass < 0 || fail < 0 || int64(pass)+int64(fail) == 0 {
			if s != nil {
				t.Fatalf("expected nil for pass=%d fail=%d", pass, fail)
			}
			return
		}
		if s == nil {
			t.Fatal("expected non-nil")
		}
		if *s < 0 || *s > 100 {
			t.Fatalf("score %d out of range", *s)
		}
		want := int32(int64(pass) * 100 / (int64(pass) + int64(fail)))
		if *s != want {
			t.Fatalf("got %d want %d", *s, want)
		}
	})
}

// FuzzScore64: severity-weighted totals are int64 sums over cluster checks.
// Same invariants as score(): nil on non-positive totals, result in [0,100].
func FuzzScore64(f *testing.F) {
	f.Add(int64(0), int64(0))
	f.Add(int64(1), int64(0))
	f.Add(int64(0), int64(1))
	f.Add(int64(10), int64(5))
	f.Add(int64(-1), int64(5))
	f.Add(int64(1<<62), int64(1<<62))
	f.Fuzz(func(t *testing.T, pass, fail int64) {
		s := score64(pass, fail)
		if pass < 0 || fail < 0 || pass+fail == 0 {
			if s != nil {
				t.Fatalf("expected nil for pass=%d fail=%d", pass, fail)
			}
			return
		}
		if s == nil {
			t.Fatal("expected non-nil")
		}
		if *s < 0 || *s > 100 {
			t.Fatalf("score %d out of range", *s)
		}
		// When pass+fail can overflow int64 addition above, Go wraps; still
		// require the returned ratio uses the same arithmetic as production.
		want := int32(pass * 100 / (pass + fail))
		if *s != want {
			t.Fatalf("got %d want %d", *s, want)
		}
	})
}

// FuzzSeverityWeight: untrusted ComplianceCheckResult severity strings map to
// the product weight table and never panic.

func FuzzSeverityWeight(f *testing.F) {
	for _, seed := range []string{"", "high", "medium", "low", "unknown", "info", "HIGH", "x"} {
		f.Add(seed)
	}
	f.Fuzz(func(t *testing.T, sev string) {
		w := severityWeight(sev)
		switch sev {
		case "high":
			if w != 10 {
				t.Fatalf("high weight = %d", w)
			}
		case "medium":
			if w != 5 {
				t.Fatalf("medium weight = %d", w)
			}
		case "low":
			if w != 2 {
				t.Fatalf("low weight = %d", w)
			}
		default:
			if w != 1 {
				t.Fatalf("default weight for %q = %d", sev, w)
			}
		}
	})
}

// FuzzClampScore: hand-edited or buggy status.score must stay in [0,100] or nil
// so Status().Update cannot fail CRD admission and freeze reconciliation.

func FuzzClampScore(f *testing.F) {
	f.Add(int32(0))
	f.Add(int32(100))
	f.Add(int32(-1))
	f.Add(int32(101))
	f.Add(int32(50))
	f.Fuzz(func(t *testing.T, v int32) {
		// nil path
		if clampScore(nil) != nil {
			t.Fatal("clampScore(nil) must be nil")
		}
		got := clampScore(&v)
		if got == nil {
			t.Fatal("clampScore(non-nil) returned nil")
		}
		if *got < 0 || *got > 100 {
			t.Fatalf("clamped score %d out of [0,100]", *got)
		}
		switch {
		case v < 0:
			if *got != 0 {
				t.Fatalf("negative %d clamped to %d, want 0", v, *got)
			}
		case v > 100:
			if *got != 100 {
				t.Fatalf("over %d clamped to %d, want 100", v, *got)
			}
		default:
			if *got != v {
				t.Fatalf("in-range %d mutated to %d", v, *got)
			}
		}
	})
}

File: console-plugin/src/scoring.test.ts
import { ClusterBaseline, ComplianceCheckResult, ResultCounts } from './models';
import { HISTORY_SCORING_MODE_ANN, aggregateCounts, checkSeverity, clusterScore, effectiveScoringMode, flatProfileScore, historyScoringModeMismatch, profileScore, scoreColor, scoreLabelColor, scoreStatus, severityWeight } from './scoring';
import { isFiniteNumber } from './parse';

// Runtime pins for fuzz sweeps: totals must be real numbers and mode checks
// real booleans whatever garbage the persisted CR carries.
const isNum = (v: unknown): v is number => typeof v === 'number';
const isBool = (v: unknown): v is boolean => typeof v === 'boolean';
import { fuzzRand, randomString } from './testing/fuzz';

describe('scoreStatus', () => {
  it.each([
    [undefined, 'danger'],
    [0, 'danger'],
    [59, 'danger'],
    [60, 'warning'],
    [89, 'warning'],
    [90, 'success'],
    [100, 'success'],
    [Number.NaN, 'danger'],
  ])('score %p -> %s', (score, status) => {
    expect(scoreStatus(score)).toBe(status);
  });
});

describe('scoreColor', () => {
  it.each([
    [undefined, 'danger'],
    [0, 'danger'],
    [59, 'danger'],
    [60, 'warning'],
    [89, 'warning'],
    [90, 'success'],
    [100, 'success'],
    // NaN must not fall through Math comparisons as a false success color.
    [Number.NaN, 'danger'],
  ])('score %p -> %s', (score, status) => {
    expect(scoreColor(score)).toContain(`status--${status}`);
  });
  it('fuzz: always a CSS var token', () => {
    let wrong: string | undefined;
    for (let i = 0; i < 500; i++) {
      const s =
        i === 0 ? undefined : i === 1 ? Number.NaN : Math.floor(fuzzRand() * 200) - 50;
      const color = scoreColor(s);
      expect(color.startsWith('var(--pf-t--')).toBeTruthy();
      // Undefined, NaN, and anything below 60 must paint danger (NaN must not
      // fall through the Math comparisons as a false success color).
      if ((s === undefined || Number.isNaN(s) || s < 60) && !color.includes('status--danger')) {
        wrong ??= `score ${String(s)} rendered ${color}, not danger`;
      }
    }
    expect(wrong).toBeUndefined();
  });
});

describe('scoreLabelColor', () => {
  it.each([
    [0, 'red'],
    [59, 'red'],
    [60, 'orange'],
    [89, 'orange'],
    [90, 'green'],
    [100, 'green'],
    // NaN comparisons are false: must not paint green/orange (same band as scoreColor danger).
    [Number.NaN, 'red'],
    // Non-finite extremes: only >=90 is green; -Infinity is red, +Infinity is green.
    [Number.NEGATIVE_INFINITY, 'red'],
    [Number.POSITIVE_INFINITY, 'green'],
  ])('score %p -> %s', (score, color) => {
    expect(scoreLabelColor(score)).toBe(color);
  });
});

describe('effectiveScoringMode / historyScoringModeMismatch', () => {
  it('defaults to Flat when scoring mode is unset', () => {
    expect(effectiveScoringMode(undefined)).toBe('Flat');
    expect(effectiveScoringMode({ spec: { profiles: ['cis'] } })).toBe('Flat');
    expect(
      effectiveScoringMode({ spec: { profiles: ['cis'], scoring: { mode: 'Flat' } } }),
    ).toBe('Flat');
    expect(
      effectiveScoringMode({
        spec: { profiles: ['cis'], scoring: { mode: 'SeverityWeighted' } },
      }),
    ).toBe('SeverityWeighted');
  });

  // Annotation is hand-editable CR metadata; unknown stamps and modes must not throw.
  it('fuzz: historyScoringModeMismatch never throws; empty stamp is not a mismatch', () => {
    let wrong: string | undefined;
    for (let i = 0; i < 500; i++) {
      const stamp =
        i % 4 === 0 ? undefined : i % 4 === 1 ? '' : i % 4 === 2 ? 'Flat' : randomString(i % 16);
      const mode: 'Flat' | 'SeverityWeighted' | undefined =
        i % 3 === 0 ? undefined : i % 3 === 1 ? 'Flat' : 'SeverityWeighted';
      const baseline = {
        metadata: {
          name: 'cluster',
          annotations: stamp === undefined ? undefined : { [HISTORY_SCORING_MODE_ANN]: stamp },
        },
        spec: {
          profiles: ['cis' as const],
          scoring: mode ? { mode } : undefined,
        },
      };
      const mismatch = historyScoringModeMismatch(baseline);
      expect(isBool(mismatch)).toBeTruthy();
      if (!stamp && mismatch) {
        wrong ??= `iteration ${i}: empty/absent stamp flagged as a mismatch`;
      }
      // effectiveScoringMode collapses anything except SeverityWeighted to Flat.
      const effective = effectiveScoringMode(baseline);
      expect(effective === 'Flat' || effective === 'SeverityWeighted').toBeTruthy();
    }
    expect(wrong).toBeUndefined();
  });

  it('detects history points stamped under a different scoring mode', () => {
    expect(historyScoringModeMismatch(undefined)).toBeFalsy();
    expect(
      historyScoringModeMismatch({
        metadata: { name: 'cluster' },
        spec: { profiles: ['cis'] },
      }),
    ).toBeFalsy();
    expect(
      historyScoringModeMismatch({
        metadata: {
          name: 'cluster',
          annotations: { [HISTORY_SCORING_MODE_ANN]: 'Flat' },
        },
        spec: { profiles: ['cis'], scoring: { mode: 'Flat' } },
      }),
    ).toBeFalsy();
    expect(
      historyScoringModeMismatch({
        metadata: {
          name: 'cluster',
          annotations: { [HISTORY_SCORING_MODE_ANN]: 'Flat' },
        },
        spec: { profiles: ['cis'], scoring: { mode: 'SeverityWeighted' } },
      }),
    ).toBeTruthy();
    expect(
      historyScoringModeMismatch({
        metadata: {
          name: 'cluster',
          annotations: { [HISTORY_SCORING_MODE_ANN]: 'SeverityWeighted' },
        },
        spec: { profiles: ['cis'] },
      }),
    ).toBeTruthy();
  });
});

describe('severityWeight / profileScore', () => {
  const check = (
    name: string,
    suite: string,
    status: ComplianceCheckResult['status'],
    severity: string,
  ): ComplianceCheckResult => ({
    metadata: {
      name,
      namespace: 'openshift-compliance',
      labels: { 'compliance.openshift.io/suite': suite },
    },
    status,
    severity,
  });

  it('matches the operator weight table', () => {
    expect(severityWeight('high')).toBe(10);
    expect(severityWeight('medium')).toBe(5);
    expect(severityWeight('low')).toBe(2);
    expect(severityWeight('unknown')).toBe(1);
    expect(severityWeight(undefined)).toBe(1);
    // Case-sensitive lockstep with operator: unexpected casing is weight 1.
    expect(severityWeight('HIGH')).toBe(1);
    expect(severityWeight('info')).toBe(1);
  });
  // Typed field wins; check-severity label is fallback; missing both is "unknown"
  // so Results severity filters and CSV match the weight table / TEST-PLAN.
  it('checkSeverity prefers .severity and falls back to the label', () => {
    expect(checkSeverity({ severity: 'high' })).toBe('high');
    expect(
      checkSeverity({
        severity: '',
        metadata: { labels: { 'compliance.openshift.io/check-severity': 'medium' } },
      }),
    ).toBe('medium');
    expect(
      checkSeverity({
        severity: 'high',
        metadata: { labels: { 'compliance.openshift.io/check-severity': 'low' } },
      }),
    ).toBe('high');
    expect(checkSeverity({})).toBe('unknown');
    expect(checkSeverity({ metadata: { labels: {} } })).toBe('unknown');
    expect(checkSeverity({ severity: '', metadata: { labels: {} } })).toBe('unknown');
  });
  it('profileScore SeverityWeighted uses label severity when field is absent', () => {
    const results: ComplianceCheckResult[] = [
      {
        metadata: {
          name: 'p1',
          namespace: 'openshift-compliance',
          labels: {
            'compliance.openshift.io/suite': 'baseline-cis',
            'compliance.openshift.io/check-severity': 'high',
          },
        },
        status: 'PASS',
      },
      {
        metadata: {
          name: 'f1',
          namespace: 'openshift-compliance',
          labels: {
            'compliance.openshift.io/suite': 'baseline-cis',
            'compliance.openshift.io/check-severity': 'low',
          },
        },
        status: 'FAIL',
      },
    ];
    // high PASS (10) + low FAIL (2) => 83; weight-1 defaults would yield 50.
    expect(
      profileScore(
        { pass: 1, fail: 1 },
        {
          mode: 'SeverityWeighted',
          filterKey: 'cis',
          results,
          profiles: ['cis'],
        },
      ),
    ).toBe(83);
  });
  it('flatProfileScore floors pass/(pass+fail)', () => {
    expect(flatProfileScore(1, 1)).toBe(50);
    expect(flatProfileScore(3, 1)).toBe(75);
    expect(flatProfileScore(0, 0)).toBeNull();
    // Lockstep with operator score(): integer division floors so a single FAIL
    // among many PASS never rounds up to a false 100.
    expect(flatProfileScore(999, 1)).toBe(99);
    expect(flatProfileScore(1, 2)).toBe(33);
    expect(flatProfileScore(1, 0)).toBe(100);
    expect(flatProfileScore(0, 5)).toBe(0);
    // Lockstep with operator score(): negative or non-finite mass is nil, not
    // a negative/NaN badge (denom>0 alone was false confidence).
    expect(flatProfileScore(-1, 5)).toBeNull();
    expect(flatProfileScore(5, -1)).toBeNull();
    expect(flatProfileScore(-1, -1)).toBeNull();
    expect(flatProfileScore(Number.NaN, 1)).toBeNull();
    expect(flatProfileScore(1, Number.POSITIVE_INFINITY)).toBeNull();
    // Finite operands whose sum or p*100 product overflows must fail closed
    // like the operator's int64 score() (nil), never NaN / Infinity: those
    // leak past threshold comparisons as a false badge color.
    expect(flatProfileScore(1e308, 1e308)).toBeNull();
    expect(flatProfileScore(Number.MAX_VALUE, Number.MAX_VALUE)).toBeNull();
    expect(flatProfileScore(1e307, 1)).toBeNull();
    expect(flatProfileScore(-1e307, -1e307)).toBeNull();
  });
  // Untrusted / hand-edited ResultCounts: score is null or in [0,100], never NaN.
  // Oracle matches operator score(): floor(pass*100/(pass+fail)) for non-neg finite.
  it('fuzz: flatProfileScore is null or [0,100] for arbitrary mass', () => {
    const samples: Array<[number | undefined, number | undefined]> = [
      [0, 0],
      [1, 0],
      [0, 1],
      [2, 1],
      [-1, 5],
      [5, -1],
      [Number.NaN, 1],
      [1, Number.POSITIVE_INFINITY],
      [Number.MAX_SAFE_INTEGER, 1],
      [Number.MAX_VALUE, Number.MAX_VALUE],
      [1e307, 1],
      [1e308, 1e308],
      [undefined, undefined],
    ];
    for (let i = 0; i < 2000; i++) {
      const pass =
        i < samples.length ? samples[i][0] : Math.floor(fuzzRand() * 1e6) - 1e3;
      const fail =
        i < samples.length ? samples[i][1] : Math.floor(fuzzRand() * 1e6) - 1e3;
      let got: number | null = null;
      expect(() => {
        got = flatProfileScore(pass, fail);
      }).not.toThrow();
      if (got === null) {
        continue;
      }
      expect(Number.isFinite(got)).toBeTruthy();
      expect(got).toBeGreaterThanOrEqual(0);
      expect(got).toBeLessThanOrEqual(100);
      const p = pass ?? 0;
      const f = fail ?? 0;
      // Oracle matches operator score(): floor(pass*100/(pass+fail)) for
      // non-negative finite mass; anything else stays unasserted beyond the
      // null/range checks above.
      const oracle =
        Number.isFinite(p) && Number.isFinite(f) && p >= 0 && f >= 0 && p + f > 0
          ? Math.floor((p * 100) / (p + f))
          : got;
      expect(got).toBe(oracle);
    }
  });
  it('profileScore uses flat counts by default', () => {
    expect(profileScore({ pass: 1, fail: 1 })).toBe(50);
  });
  // SeverityWeighted with an empty CCR list (watch still loading) must not blank
  // the Overview badge when status already has pass/fail tallies.
  it('profileScore SeverityWeighted falls back to flat when results are empty', () => {
    expect(
      profileScore(
        { pass: 3, fail: 1 },
        {
          mode: 'SeverityWeighted',
          filterKey: 'cis',
          results: [],
          profiles: ['cis'],
        },
      ),
    ).toBe(75);
    // Prefer last history point (operator weighted) over flat when present.
    expect(
      profileScore(
        { pass: 1, fail: 1 },
        {
          mode: 'SeverityWeighted',
          filterKey: 'cis',
          results: [],
          profiles: ['cis'],
          history: [{ score: 83 }, { score: 90 }],
        },
      ),
    ).toBe(90);
    // Loaded path with no countable mass still null (not a false flat score).
    expect(
      profileScore(
        { pass: 0, fail: 0 },
        {
          mode: 'SeverityWeighted',
          filterKey: 'cis',
          results: [check('m1', 'baseline-cis', 'MANUAL', 'medium')],
          profiles: ['cis'],
        },
      ),
    ).toBeNull();
  });
  it('profileScore weights by severity in SeverityWeighted mode', () => {
    const results = [
      check('p1', 'baseline-cis', 'PASS', 'high'),
      check('f1', 'baseline-cis', 'FAIL', 'low'),
    ];
    // high PASS (10) + low FAIL (2) => 83; flat would be 50
    expect(
      profileScore(
        { pass: 1, fail: 1 },
        {
          mode: 'SeverityWeighted',
          filterKey: 'cis',
          results,
          profiles: ['cis'],
        },
      ),
    ).toBe(83);
    expect(
      profileScore(
        { pass: 1, fail: 1 },
        { mode: 'Flat', filterKey: 'cis', results, profiles: ['cis'] },
      ),
    ).toBe(50);
  });
  // Overview cards recompute severity-weighted scores client-side; a waived FAIL
  // must leave the denominator so the badge matches status.score.
  it('profileScore SeverityWeighted excludes waived FAILs', () => {
    const results = [
      check('p1', 'baseline-cis', 'PASS', 'high'),
      check('f1', 'baseline-cis', 'FAIL', 'high'),
    ];
    const now = new Date('2026-07-11T00:00:00Z');
    // Without waiver: high PASS (10) + high FAIL (10) => 50
    expect(
      profileScore(
        { pass: 1, fail: 1 },
        {
          mode: 'SeverityWeighted',
          filterKey: 'cis',
          results,
          profiles: ['cis'],
          waivers: [{ name: 'f1', reason: 'accepted' }],
          now,
        },
      ),
    ).toBe(100);
    // Expired waiver must re-include the FAIL.
    expect(
      profileScore(
        { pass: 1, fail: 1 },
        {
          mode: 'SeverityWeighted',
          filterKey: 'cis',
          results,
          profiles: ['cis'],
          waivers: [{ name: 'f1', expiresAt: '2026-07-10T00:00:00Z' }],
          now,
        },
      ),
    ).toBe(50);
    // Only FAILs, all waived: no countable mass => null (not a false 100/0).
    expect(
      profileScore(
        { pass: 0, fail: 2 },
        {
          mode: 'SeverityWeighted',
          filterKey: 'cis',
          results: [
            check('f1', 'baseline-cis', 'FAIL', 'high'),
            check('f2', 'baseline-cis', 'FAIL', 'low'),
          ],
          profiles: ['cis'],
          waivers: [{ name: 'f1' }, { name: 'f2' }],
          now,
        },
      ),
    ).toBeNull();
    // Foreign / stale waiver names must not invent matches (by-name contract).
    expect(
      profileScore(
        { pass: 1, fail: 1 },
        {
          mode: 'SeverityWeighted',
          filterKey: 'cis',
          results,
          profiles: ['cis'],
          waivers: [{ name: 'not-a-real-check' }, { name: '' }],
          now,
        },
      ),
    ).toBe(50);
    // Shared activeWaived Set path (Overview prebuilds one Set for all cards).
    const activeWaived = new Set(['f1']);
    expect(
      profileScore(
        { pass: 1, fail: 1 },
        {
          mode: 'SeverityWeighted',
          filterKey: 'cis',
          results,
          profiles: ['cis'],
          activeWaived,
          now,
        },
      ),
    ).toBe(100);
  });
  // Overview pre-buckets by suite and ownership, then omits profiles so
  // profileScore weighs the bucket without a second membership scan.
  it('profileScore SeverityWeighted prefiltered bucket skips ownership re-scan', () => {
    const bucket = [
      check('p1', 'baseline-cis', 'PASS', 'high'),
      check('f1', 'baseline-cis', 'FAIL', 'low'),
    ];
    // high PASS (10) + low FAIL (2) => 83; no profiles/tailored => trust bucket.
    expect(
      profileScore(
        { pass: 1, fail: 1 },
        {
          mode: 'SeverityWeighted',
          filterKey: 'cis',
          results: bucket,
        },
      ),
    ).toBe(83);
  });
  // Multi-profile watches return every suite; score for one card must ignore
  // foreign suites and unselected tailored results.
  it('profileScore SeverityWeighted filters by suite and ownership', () => {
    const results = [
      check('cis-pass', 'baseline-cis', 'PASS', 'high'),
      check('stig-fail', 'baseline-stig', 'FAIL', 'high'),
      check('tp-fail', 'baseline-tp-custom', 'FAIL', 'high'),
    ];
    expect(
      profileScore(
        { pass: 1, fail: 0 },
        {
          mode: 'SeverityWeighted',
          filterKey: 'cis',
          results,
          profiles: ['cis'],
          tailoredProfiles: ['custom'],
        },
      ),
    ).toBe(100);
    expect(
      profileScore(
        { pass: 0, fail: 1 },
        {
          mode: 'SeverityWeighted',
          filterKey: 'tp-custom',
          results,
          profiles: ['cis'],
          tailoredProfiles: ['custom'],
        },
      ),
    ).toBe(0);
    // Tailored-only baseline: profiles may be empty; still recompute weights.
    expect(
      profileScore(
        { pass: 0, fail: 1 },
        {
          mode: 'SeverityWeighted',
          filterKey: 'tp-custom',
          results,
          profiles: [],
          tailoredProfiles: ['custom'],
        },
      ),
    ).toBe(0);
  });
});

describe('aggregateCounts', () => {
  const c = (
    pass: number,
    fail: number,
    manual = 0,
    info = 0,
    error = 0,
    inconsistent = 0,
    waived = 0,
    notApplicable = 0,
  ) => ({
    pass,
    fail,
    manual,
    info,
    error,
    inconsistent,
    waived,
    notApplicable,
  });
  it('sums profiles and tailored profiles together', () => {
    expect(aggregateCounts(c(10, 2, 1, 4, 0, 5, 7), c(40, 8, 3, 1, 0, 6, 2))).toEqual(
      c(50, 10, 4, 5, 0, 11, 9),
    );
  });
  it('returns zeros for no groups', () => {
    expect(aggregateCounts()).toEqual(c(0, 0, 0, 0, 0, 0, 0, 0));
  });
  it('score composition matches: tailored-only results still populate totals', () => {
    // regular profile empty, tailored has results -> totals non-zero
    const totals = aggregateCounts(c(0, 0), c(2, 1));
    expect(totals.pass + totals.fail).toBe(3);
  });
  it('treats missing count fields from older persisted status as zero', () => {
    // SAFETY: older persisted status omits the newer count fields on purpose;
    // aggregateCounts must default them to zero, not propagate undefined.
    const totals = aggregateCounts({ pass: 1, fail: 2 } as ResultCounts);
    expect(totals).toEqual(c(1, 2, 0, 0, 0, 0, 0, 0));
  });
  it('folds a non-finite (NaN/Infinity) or non-numeric count to 0, never poisoning totals', () => {
    // A tampered non-numeric/non-finite count must not string-concatenate or
    // spread NaN/Infinity across the donut totals.
    // SAFETY: stale persisted status may carry non-finite counts; the fold must
    // keep totals finite.
    const nonFinite = { pass: Number.NaN, fail: Number.POSITIVE_INFINITY } as ResultCounts;
    // SAFETY: tampered persisted status carrying a numeric-string count ('5');
    // aggregation must coerce it instead of concatenating or throwing.
    const stringy = JSON.parse('{"pass": "5", "fail": 3}') as ResultCounts;
    const totals = aggregateCounts(nonFinite, stringy);
    expect(totals.pass).toBe(5); // NaN -> 0, '5' -> 5
    expect(totals.fail).toBe(3); // Infinity -> 0, 3 -> 3
    expect(Number.isFinite(totals.pass + totals.fail)).toBeTruthy();
  });
  it('keeps running totals finite when huge-but-finite counts overflow the accumulator', () => {
    // Per-value folding is not enough: two finite 9e307 counts sum to Infinity.
    // The accumulator must saturate (keep the prior finite total) so donut
    // segments and the report total never receive a non-finite y value.
    // SAFETY: stale CR status omits the newer count fields; huge-but-finite
    // contributions must saturate instead of overflowing to Infinity.
    const first = { pass: 9e307, fail: -9e307 } as ResultCounts;
    // SAFETY: same partial persisted shape; both halves must saturate.
    const second = { pass: 9e307, fail: -9e307 } as ResultCounts;
    const totals = aggregateCounts(first, second);
    expect(Number.isFinite(totals.pass)).toBeTruthy();
    expect(Number.isFinite(totals.fail)).toBeTruthy();
    // Saturation keeps the first contribution rather than inventing Infinity.
    expect(totals.pass).toBe(9e307);
    expect(totals.fail).toBe(-9e307);
  });
  // Status count fields may be missing, huge, or negative from stale CRs.
  it('fuzz: never throws; all fields are finite numbers', () => {
    for (let i = 0; i < 500; i++) {
      const partial = () => ({
        pass: i % 7 === 0 ? undefined : (i % 1000) - 50,
        fail: i % 5 === 0 ? undefined : (i % 800) - 20,
        manual: i % 11 === 0 ? undefined : i % 30,
        info: i % 13 === 0 ? undefined : i % 40,
        error: i % 17 === 0 ? undefined : i % 10,
        inconsistent: i % 19 === 0 ? undefined : i % 15,
        waived: i % 23 === 0 ? undefined : i % 25,
        notApplicable: i % 29 === 0 ? undefined : i % 12,
      });
      // SAFETY: stale CR status carries undefined/garbage count fields; the
      // aggregation must default them so every total stays a finite number.
      const g = (): ResultCounts => partial() as ResultCounts;
      const totals = aggregateCounts(g(), g(), g());
      for (const v of Object.values(totals)) {
        expect(isFiniteNumber(v)).toBeTruthy();
      }
    }
  });
});

describe('clusterScore', () => {
  // SAFETY: minimal CR fixture; clusterScore reads only metadata.name and
  // status.score and must tolerate partial CR shapes (no spec block).
  const cb = (name: string, score?: number): ClusterBaseline =>
    ({ metadata: { name }, status: score == null ? {} : { score } }) as ClusterBaseline;

  it('returns null when there is no baseline', () => {
    expect(clusterScore(undefined)).toBeNull();
    expect(clusterScore([])).toBeNull();
  });
  it('prefers the singleton named "cluster"', () => {
    expect(clusterScore([cb('other', 10), cb('cluster', 95)])).toBe(95);
  });
  it('ignores objects not named "cluster"', () => {
    expect(clusterScore([cb('a', 42), cb('b', 7)])).toBeNull();
  });
  it('returns null when the baseline exists but has not scored', () => {
    expect(clusterScore([cb('cluster')])).toBeNull();
  });
  it('treats a zero score as a value, not absent', () => {
    expect(clusterScore([cb('cluster', 0)])).toBe(0);
  });
  // Baseline list shape is untrusted API data; only the singleton counts.
  it('fuzz: never throws; null or a number', () => {
    for (let i = 0; i < 500; i++) {
      const list: ClusterBaseline[] = Array.from({ length: i % 5 }, (_, j) =>
        cb(j === 1 ? 'cluster' : randomString((j + 1) % 12), i % 3 === 0 ? undefined : (i + j) % 101),
      );
      const got = clusterScore(i % 7 === 0 ? undefined : list);
      expect(got === null || isNum(got)).toBeTruthy();
    }
  });
});
